pkg/operator/ceph/cluster/osd/create.go

```diff
@@ -0,0 +1,391 @@
+		if walOK {
+			logger.Infof("OSD will have its wal device on %q", walSource.ClaimName)
+		}
+
+		osdProps := osdProperties{
+			crushHostname:    dataSource.ClaimName,
+			pvc:              dataSource,
+			metadataPVC:      metadataSource,
+			walPVC:           walSource,
+			resources:        volume.Resources,
+			placement:        volume.Placement,
+			preparePlacement: volume.PreparePlacement,
+			portable:         volume.Portable,
+			schedulerName:    volume.SchedulerName,
+			encrypted:        volume.Encrypted,
+			deviceSetName:    volume.Name,
+		}
+		osdProps.storeConfig.DeviceClass = volume.CrushDeviceClass
+
+		// TODO: was this left from some personal debugging?
+		logger.Debugf("osdProps are %+v", osdProps)
+
+		if osdProps.encrypted {
+			// If the deviceSet template has "encrypted" but the Ceph version is not compatible
+			if !c.isCephVolumeRawModeSupported() {
+				errMsg := fmt.Sprintf("failed to validate storageClassDeviceSet %q. min required ceph version to support encryption is %q or %q", volume.Name, cephVolumeRawEncryptionModeMinNautilusCephVersion.String(), cephVolumeRawEncryptionModeMinOctopusCephVersion.String())
+				errs.addError(errMsg)
+				continue
+			}
+
+			// create encryption Kubernetes Secret if the PVC is encrypted
+			key, err := generateDmCryptKey()
+			if err != nil {
+				errMsg := fmt.Sprintf("failed to generate dmcrypt key for osd claim %q. %v", osdProps.pvc.ClaimName, err)
+				errs.addError(errMsg)
+				continue
+			}
+
+			// Initialize the KMS code
+			kmsConfig := kms.NewConfig(c.context, &c.spec, c.clusterInfo)
+
+			// We could set an env var in the Operator or a global var instead of the API call?
+			// Hopefully, the API is cheap and we can always retrieve the token if it has changed...
+			if c.spec.Security.KeyManagementService.IsTokenAuthEnabled() {
+				err := kms.SetTokenToEnvVar(c.context, c.spec.Security.KeyManagementService.TokenSecretName, kmsConfig.Provider, c.clusterInfo.Namespace)
+				if err != nil {
+					errMsg := fmt.Sprintf("failed to fetch kms token secret %q. %v", c.spec.Security.KeyManagementService.TokenSecretName, err)
+					errs.addError(errMsg)
+					continue
+				}
+			}
+
+			// Generate and store the encrypted key in whatever KMS is configured
+			err = kmsConfig.PutSecret(osdProps.pvc.ClaimName, key)
+			if err != nil {
+				errMsg := fmt.Sprintf("failed to store secret. %v", err)
+				errs.addError(errMsg)
+				continue
+			}
+		}
+
+		// Skip OSD prepare if deployment already exists for the PVC
+		// NOTE: this used to add OrchestrationStatusAlreadyCompleted for the PVC (before that
+		// it added OrchestrationStatusCompleted), but I changed the behavior to not create a
+		// status since status is really only used for new creations and not updates
+		if existingDeployments.Contains(dataSource.ClaimName) {
+			logger.Infof("skipping OSD prepare pod creation for PVC %q because OSD daemon using the PVC already exists", osdProps.crushHostname)
+			continue
+		}
```

Comment:
Thoughts @travisn @leseb ?

Reply:
How does the OSD get upgraded? If you implemented the OSD upgrade separately, sounds good not to create the OrchestrationStatusCompleted.

Reply:
OSD update/upgrade is separate from creation, so my thinking is that prepare jobs only get created (and counted) when the jobs are running and have the possibility of actually creating new OSDs.